Microsoft.Azure.Cosmos/src/Routing/PartitionKeyHashRanges.cs

```
@@ -0,0 +1,212 @@
+
+        private PartitionKeyHashRanges(SortedSet<PartitionKeyHashRange> partitionKeyHashRanges)
+        {
+            // All invariants are checked by the static constructor
+            this.partitionKeyHashRanges = partitionKeyHashRanges;
+        }
+
+        public static PartitionKeyHashRanges Create(IEnumerable<PartitionKeyHashRange> partitionKeyHashRanges)
+        {
+            CreateOutcome createStatus = PartitionKeyHashRanges.TryCreate(
+                partitionKeyHashRanges,
+                out PartitionKeyHashRanges partitionedSortedEffectiveRanges);
+
+            switch (createStatus)
+            {
+                case CreateOutcome.DuplicatePartitionKeyRange:
+                    throw new ArgumentException($"{nameof(partitionKeyHashRanges)} must not have duplicate values.");
+
+                case CreateOutcome.EmptyPartitionKeyRange:
+                    throw new ArgumentException($"{nameof(partitionKeyHashRanges)} must not have an empty range.");
+
+                case CreateOutcome.NoPartitionKeyRanges:
+                    throw new ArgumentException($"{nameof(partitionKeyHashRanges)} must not be empty.");
+
+                case CreateOutcome.NullPartitionKeyRanges:
+                    throw new ArgumentNullException(nameof(partitionKeyHashRanges));
+
+                case CreateOutcome.RangesAreNotContiguous:
+                    throw new ArgumentException($"{nameof(partitionKeyHashRanges)} must have contiguous ranges.");
+
+                case CreateOutcome.RangesOverlap:
+                    throw new ArgumentException($"{nameof(partitionKeyHashRanges)} must not overlapping ranges.");
+
+                case CreateOutcome.Success:
+                    return partitionedSortedEffectiveRanges;
+
+                default:
+                    throw new ArgumentOutOfRangeException($"Unknown {nameof(CreateOutcome)}: {createStatus}.");
+            }
+        }
+
+        public static CreateOutcome TryCreate(
```

Why not just return an exception instead of having another enum type?

---

You can't switch on a regular exception and it creates an allocation. The user is free to use the non Try version of this method.